lib/Sema/TypeCheckDeclPrimary.cpp

```diff
@@ -3670,40 +3670,114 @@ void TypeChecker::checkParameterList(ParameterList *params,
   auto *dc = MED->getDeclContext();
-  auto foundMacros = TypeChecker::lookupMacros(
-      MED->getDeclContext(), MED->getMacro(),
-      MED->getLoc(), MacroRole::Declaration);
+  // Non-call declaration macros cannot be overloaded.
+  MacroRoles viableRoles(MacroRole::Declaration);
+  // In a code block context, a macro expansion decl can also expand to an
+  // arbitrary code item.
+  if (dc->isLocalContext() ||
+      dc->getContextKind() == DeclContextKind::FileUnit) {
+    viableRoles |= MacroRole::Expression;
+    viableRoles |= MacroRole::CodeItem;
+  }
+  SmallVector<MacroDecl *, 4> foundMacros;
+  auto foundRoles = TypeChecker::lookupMacros(
+      MED->getDeclContext(), MED->getMacro(), MED->getLoc(), viableRoles,
+      foundMacros);
   if (foundMacros.empty()) {
     MED->diagnose(diag::macro_undefined, MED->getMacro().getBaseIdentifier())
         .highlight(MED->getMacroLoc().getSourceRange());
-    return {};
+    return nullptr;
   }
+
+  auto expandAsMacroExpansionExpr = [&]() -> BraceStmt * {
+    Expr *mee = MED->createExpr();
+    TypeChecker::typeCheckExpression(mee, dc, {},
+                                     TypeCheckExprFlags::IsExprStmt |
+                                         TypeCheckExprFlags::IsDiscarded);
+    ASTNode expansion = mee;
+    // If top-level, wrap it in a `TopLevelCodeDecl(BraceStmt(...))`.
+    auto innermostDC = dc;
+    if (dc->getContextKind() == DeclContextKind::FileUnit) {
+      auto *tlcd = new (ctx) TopLevelCodeDecl(
+          dc, BraceStmt::createImplicit(ctx, {expansion}));
+      tlcd->setDeclContext(dc);
+      expansion = tlcd;
+      innermostDC = tlcd;
+    }
+    auto codeBlockItemList = BraceStmt::createImplicit(ctx, {expansion});
+    MED->setRewritten(codeBlockItemList);
+    TypeChecker::postTypeCheckMacroExpansion(mee, innermostDC);
+    return codeBlockItemList;
+  };
+
+  // If we've only found expression macros, expand to a macro expansion expr.
+  if (foundRoles.containsOnly(MacroRole::Expression))
+    return expandAsMacroExpansionExpr();
+
   // Resolve macro candidates.
   MacroDecl *macro;
-  // Non-call declaration macros cannot be overloaded.
+  // Non-call macros cannot be overloaded.
   auto *args = MED->getArgs();
   if (!args) {
     if (foundMacros.size() > 1) {
       MED->diagnose(diag::ambiguous_decl_ref, MED->getMacro())
           .highlight(MED->getMacroLoc().getSourceRange());
       for (auto *candidate : foundMacros)
         candidate->diagnose(diag::found_candidate);
-      return {};
+      return nullptr;
     }
     macro = foundMacros.front();
   }
     // Call-like macros need to be resolved.
   else {
     using namespace constraints;
     ConstraintSystem cs(dc, ConstraintSystemOptions());
-    // Type-check macro arguments.
-    for (auto *arg : args->getArgExprs())
-      cs.setType(arg, TypeChecker::typeCheckExpression(arg, dc));
+    SmallVector<Argument, 2> newArgs;
+    newArgs.reserve(args->size());
+    for (auto arg : *args) {
```

Comment:
There's similar logic to this for attached macros, where we need to type-check `@macroname(args...)`. We should unify the logic with that.

Reply:
Fixed in #63247.  `ResolveAttachedMacroRequest`'s approach of forming and type-checking a `OverloadedDeclRefExpr` was very clever. I was able to remove all of the custom constraint generation code.